File: packages/contracts/go/table.go
// Table contracts — Milestone 1 (ADR-011). Mirrors src/types/table.ts.
//
// Authority split (§50.1, no split-authority columns):
//   RestaurantTable — pure config, cloud→edge, versioned, replaced wholesale.
//   TableSession    — operational, edge→cloud, append-only replay.
package contracts

import "time"

type RestaurantTable struct {
	ID            string `json:"id"`
	OutletID      string `json:"outlet_id"`
	Section       string `json:"section"`
	Label         string `json:"label"`
	SeatCount     int    `json:"seat_count"`
	IsActive      bool   `json:"is_active"`
	ConfigVersion int    `json:"config_version"`
	SchemaVersion int    `json:"schema_version"`
}

type TableSessionState string

// Stored session states. AVAILABLE is not stored — a table with no open
// session is available (see TableDisplayState).
const (
	TableSessionStateOccupied       TableSessionState = "OCCUPIED"
	TableSessionStateOrdered        TableSessionState = "ORDERED"
	TableSessionStateKotSent        TableSessionState = "KOT_SENT"
	TableSessionStateFoodReady      TableSessionState = "FOOD_READY"
	TableSessionStateBillRequested  TableSessionState = "BILL_REQUESTED"
	TableSessionStatePaymentPending TableSessionState = "PAYMENT_PENDING"
	TableSessionStatePaid           TableSessionState = "PAID"
	TableSessionStateDirty          TableSessionState = "DIRTY"
	TableSessionStateClosed         TableSessionState = "CLOSED"
)

type TableDisplayState string

// Floor-plan states docs/spec/tables.md renders. RESERVED is spec-defined but
// nothing in Milestone 1 produces it (reservations are Milestone 9).
const (
	TableDisplayStateAvailable TableDisplayState = "AVAILABLE"
	TableDisplayStateReserved  TableDisplayState = "RESERVED"
)

type TableSession struct {
	ID              string            `json:"id"`
	OutletID        string            `json:"outlet_id"`
	TableID         string            `json:"table_id"`
	State           TableSessionState `json:"state"`
	CurrentOrderID  *string           `json:"current_order_id"`
	GuestCount      int               `json:"guest_count"`
	OpenedByUserID  *string           `json:"opened_by_user_id"`
	OpenedAt        time.Time         `json:"opened_at"`
	ClosedAt        *time.Time        `json:"closed_at"`
	Version         int               `json:"version"`
	CreatedAt       time.Time         `json:"created_at"`
	UpdatedAt       time.Time         `json:"updated_at"`
	SchemaVersion   int               `json:"schema_version"`
}

File: packages/contracts/src/types/table.ts
// Table contracts — Milestone 1 (ADR-011). Mirrors go/table.go.
//
// Authority split (§50.1, no split-authority columns):
//   RestaurantTable  — pure config, cloud→edge, versioned, replaced wholesale.
//   TableSession     — operational, edge→cloud, append-only replay.

import { z } from "zod";

export const RestaurantTableSchema = z.object({
  id: z.string().uuid(),
  outlet_id: z.string().uuid(),
  section: z.string(), // floor / zone, e.g. 'GROUND', 'TERRACE'
  label: z.string(), // 'T4', 'G12'
  seat_count: z.number().int().positive(),
  is_active: z.boolean(),
  config_version: z.number().int(),
  schema_version: z.literal(1),
});
export type RestaurantTable = z.infer<typeof RestaurantTableSchema>;

// Stored state of an open seating. AVAILABLE is not a stored value — a table
// with no open session IS available (see TableDisplayState below).
export const TableSessionStateSchema = z.enum([
  "OCCUPIED",
  "ORDERED",
  "KOT_SENT",
  "FOOD_READY",
  "BILL_REQUESTED",
  "PAYMENT_PENDING",
  "PAID",
  "DIRTY",
  "CLOSED",
]);
export type TableSessionState = z.infer<typeof TableSessionStateSchema>;

// The floor-plan state docs/spec/tables.md renders: the open session's state,
// or AVAILABLE when there is none. RESERVED is defined by the spec but is not
// produced until reservations land (Milestone 9) — nothing in Milestone 1
// writes it.
export const TableDisplayStateSchema = z.enum([
  "AVAILABLE",
  "RESERVED",
  ...TableSessionStateSchema.options,
]);
export type TableDisplayState = z.infer<typeof TableDisplayStateSchema>;

export const TableSessionSchema = z.object({
  id: z.string().uuid(),
  outlet_id: z.string().uuid(),
  table_id: z.string().uuid(),
  state: TableSessionStateSchema,
  current_order_id: z.string().uuid().nullable(),
  guest_count: z.number().int().positive(),
  opened_by_user_id: z.string().uuid().nullable(),
  opened_at: z.string().datetime(),
  closed_at: z.string().datetime().nullable(),
  version: z.number().int(), // optimistic concurrency, sync envelope field
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  schema_version: z.literal(1),
});
export type TableSession = z.infer<typeof TableSessionSchema>;
